File: android/src/test/java/to/holepunch/bare/android/BareProtocolTest.kt
package to.holepunch.bare.android

import org.junit.Assert.assertArrayEquals
import org.junit.Assert.assertEquals
import org.junit.Assert.assertNull
import org.junit.Assert.assertTrue
import org.junit.Assert.assertThrows
import org.junit.Test
import java.nio.ByteBuffer

class BareProtocolTest {
    private val header = """{"type":"DISPATCH","pluginId":"core.health","event":"health.ping"}"""

    @Test
    fun `round-trips type header and payload`() {
        val payload = byteArrayOf(1, 2, 3)
        val buffer = BareProtocol.buildMessage(BareProtocol.MessageType.ENVELOPE, header, payload)

        val msg = BareProtocol.parseMessage(buffer)
        assertTrue(msg != null)
        assertEquals(BareProtocol.MessageType.ENVELOPE, msg!!.type)
        assertEquals(header, msg.header)
        assertArrayEquals(payload, msg.payload)
    }

    @Test
    fun `parses a null payload as null`() {
        val buffer = BareProtocol.buildMessage(BareProtocol.MessageType.ENVELOPE, header, null)
        val msg = BareProtocol.parseMessage(buffer)
        assertTrue(msg != null)
        assertNull(msg!!.payload)
    }

    @Test
    fun `rejects an unsupported version`() {
        val buffer = BareProtocol.buildMessage(BareProtocol.MessageType.ENVELOPE, header, null)
        buffer.put(0, 99.toByte())
        assertNull(BareProtocol.parseMessage(buffer))
    }

    @Test
    fun `rejects an unsupported message type`() {
        val buffer = BareProtocol.buildMessage(BareProtocol.MessageType.ENVELOPE, header, null)
        buffer.put(1, 250.toByte())
        assertNull(BareProtocol.parseMessage(buffer))
    }

    @Test
    fun `returns null for a truncated buffer`() {
        val full = BareProtocol.buildMessage(BareProtocol.MessageType.ENVELOPE, header, null)
        val bytes = ByteArray(full.remaining())
        full.get(bytes)
        val truncated = ByteBuffer.wrap(bytes.copyOf(bytes.size - 1))
        assertNull(BareProtocol.parseMessage(truncated))
    }

    @Test
    fun `rejects a header larger than the 16-bit length field`() {
        val bigHeader = """{"type":"DISPATCH","pluginId":"core.health","event":"e","args":{"x":"${"y".repeat(0x10000)}"}}"""
        assertThrows(IllegalArgumentException::class.java) {
            BareProtocol.buildMessage(BareProtocol.MessageType.ENVELOPE, bigHeader, null)
        }
    }

    @Test
    fun `rejects a frame larger than the maximum`() {
        val payload = ByteArray(BareProtocol.MAX_FRAME_BYTES)
        assertThrows(IllegalArgumentException::class.java) {
            BareProtocol.buildMessage(BareProtocol.MessageType.ENVELOPE, header, payload)
        }
    }

    @Test
    fun `returns null for an oversized frame on parse`() {
        val payload = ByteArray(BareProtocol.MAX_FRAME_BYTES + 1)
        val buffer = ByteBuffer.allocateDirect(4 + payload.size)
        buffer.put(BareProtocol.VERSION)
        buffer.put(BareProtocol.MessageType.ENVELOPE)
        buffer.putShort(payload.size.toShort())
        buffer.put(payload)
        buffer.flip()
        assertNull(BareProtocol.parseMessage(buffer))
    }

    @Test
    fun `builds an INVOKE_RESPONSE error envelope from a request header`() {
        val requestHeader = """{"type":"INVOKE_REQUEST","pluginId":"core.health","event":"health.ping","requestId":"req-1"}"""
        val err = BareProtocol.buildErrorResponse(
            requestHeader,
            ErrorCodes.FRAME_TOO_LARGE,
            "too big",
        )
        assertEquals(BareProtocol.MessageType.ENVELOPE, err.type)
        assertTrue(err.header.contains("\"requestId\":\"req-1\""))
        assertTrue(err.header.contains("\"code\":\"FRAME_TOO_LARGE\""))
        assertNull(err.payload)
    }
}

class FrameDecoderTest {
    private val header = """{"type":"DISPATCH","pluginId":"core.health","event":"health.ping"}"""

    private fun frame(payload: ByteArray? = null): ByteArray {
        val buffer = BareProtocol.buildMessage(BareProtocol.MessageType.ENVELOPE, header, payload)
        val bytes = ByteArray(buffer.remaining())
        buffer.get(bytes)
        return bytes
    }

    @Test
    fun `reassembles a frame split across chunks`() {
        val f = frame(byteArrayOf(1, 2, 3, 4))
        val decoder = FrameDecoder()
        val decoded = mutableListOf<BareProtocol.WireMessage>()
        for (i in f.indices) {
            decoded += decoder.push(f.copyOfRange(i, i + 1))
        }
        assertEquals(1, decoded.size)
        assertArrayEquals(byteArrayOf(1, 2, 3, 4), decoded[0].payload)
    }

    @Test
    fun `drains coalesced frames in order`() {
        val f1 = frame(null)
        val f2 = frame(byteArrayOf(9))
        val both = f1 + f2
        val decoder = FrameDecoder()
        val decoded = decoder.push(both)
        assertEquals(2, decoded.size)
        assertArrayEquals(byteArrayOf(9), decoded[1].payload)
    }

    @Test
    fun `skips a corrupt complete frame without desyncing the rest`() {
        val f1 = frame(null)
        val bad = frame(null).also { it[0] = 99 }
        val f2 = frame(byteArrayOf(7))
        val both = f1 + bad + f2
        val decoder = FrameDecoder()
        val decoded = decoder.push(both)
        // The corrupt frame is skipped; the two good frames still parse.
        assertEquals(2, decoded.size)
        assertArrayEquals(byteArrayOf(7), decoded[1].payload)
    }

    @Test
    fun `goes inert after a frame too large and resyncs on clear`() {
        val decoder = FrameDecoder()
        // headerLen 0, payloadLen 0xffffff: frameLen exceeds the cap (a framing
        // violation).
        val oversized = byteArrayOf(
            BareProtocol.VERSION,
            BareProtocol.MessageType.ENVELOPE,
            0, 0, 0xff.toByte(), 0xff.toByte(), 0xff.toByte(),
        )
        val first = decoder.push(oversized)
        assertEquals(0, first.size)
        assertTrue(decoder.error != null)

        decoder.clear()
        val good = decoder.push(frame(null))
        assertEquals(1, good.size)
        assertTrue(decoder.error == null)
    }
}

File: ios/Tests/BareHostTests/BareProtocolTest.swift
import XCTest
import BareHost

final class BareProtocolTest: XCTestCase {
  private let header =
    #"{"type":"DISPATCH","pluginId":"core.health","event":"health.ping"}"#

  func testRoundTripsTypeHeaderAndPayload() throws {
    let payload = Data([1, 2, 3])
    let buffer = try BareProtocol.buildMessage(
      type: BareProtocol.MessageType.envelope,
      headerJson: header,
      payload: payload
    )

    let message = BareProtocol.parseMessage(buffer)
    XCTAssertNotNil(message)
    XCTAssertEqual(BareProtocol.MessageType.envelope, message?.type)
    XCTAssertEqual(header, message?.header)
    XCTAssertEqual(payload, message?.payload)
  }

  func testParsesNilPayloadAsNil() throws {
    let buffer = try BareProtocol.buildMessage(
      type: BareProtocol.MessageType.envelope,
      headerJson: header,
      payload: nil
    )
    let message = BareProtocol.parseMessage(buffer)
    XCTAssertNotNil(message)
    XCTAssertNil(message?.payload)
  }

  func testRejectsAnUnsupportedVersion() throws {
    var buffer = try BareProtocol.buildMessage(
      type: BareProtocol.MessageType.envelope,
      headerJson: header,
      payload: nil
    )
    buffer[0] = 99
    XCTAssertNil(BareProtocol.parseMessage(buffer))
  }

  func testRejectsAnUnsupportedMessageType() throws {
    var buffer = try BareProtocol.buildMessage(
      type: BareProtocol.MessageType.envelope,
      headerJson: header,
      payload: nil
    )
    buffer[1] = 250
    XCTAssertNil(BareProtocol.parseMessage(buffer))
  }

  func testReturnsNilForATruncatedBuffer() throws {
    let full = try BareProtocol.buildMessage(
      type: BareProtocol.MessageType.envelope,
      headerJson: header,
      payload: nil
    )
    XCTAssertNil(BareProtocol.parseMessage(full.dropLast(1)))
  }

  func testRejectsASubSevenByteFrame() {
    // The envelope is [version][type][headerLen:2][payloadLen:3] = 7 bytes at
    // minimum. Shorter buffers must be rejected, not trapped by an out-of-bounds
    // read of bytes[4..6] when deriving payloadLen.
    XCTAssertNil(BareProtocol.parseMessage(Data([1, 2, 3, 4])))
    XCTAssertNil(BareProtocol.parseMessage(Data([1, 2, 3, 4, 5])))
    XCTAssertNil(BareProtocol.parseMessage(Data([1, 2, 3, 4, 5, 6])))
  }

  func testRejectsHeaderLargerThanThe16BitLengthField() {
    let bigHeader =
      #"{"type":"DISPATCH","pluginId":"core.health","event":"e","args":{"x":"# +
      String(repeating: "y", count: 0x10000) +
      #""}}"#
    XCTAssertThrowsError(
      try BareProtocol.buildMessage(
        type: BareProtocol.MessageType.envelope,
        headerJson: bigHeader,
        payload: nil
      )
    )
  }

  func testRejectsAFrameLargerThanTheMaximum() {
    let payload = Data(count: BareProtocol.maxFrameBytes)
    XCTAssertThrowsError(
      try BareProtocol.buildMessage(
        type: BareProtocol.MessageType.envelope,
        headerJson: header,
        payload: payload
      )
    )
  }

  func testReturnsNilForAnOversizedFrameOnParse() {
    var buffer = Data()
    buffer.append(BareProtocol.version)
    buffer.append(BareProtocol.MessageType.envelope)
    buffer.append(0)
    buffer.append(0)
    buffer.append(Data(count: BareProtocol.maxFrameBytes + 1))
    XCTAssertNil(BareProtocol.parseMessage(buffer))
  }

  func testBuildsAnInvokeResponseErrorEnvelopeFromARequestHeader() throws {
    let requestHeader =
      #"{"type":"INVOKE_REQUEST","pluginId":"core.health","event":"health.ping","requestId":"req-1"}"#
    let error = BareProtocol.buildErrorResponse(
      headerJson: requestHeader,
      code: ErrorCodes.frameTooLarge,
      message: "too big"
    )
    XCTAssertEqual(BareProtocol.MessageType.envelope, error.type)
    XCTAssertTrue(error.header.contains("\"requestId\":\"req-1\""))
    XCTAssertTrue(error.header.contains("\"code\":\"FRAME_TOO_LARGE\""))
    XCTAssertNil(error.payload)
  }
}
